### PR TITLE
[CI Fest] Fix `02164_clickhouse_local_interactive_exception`

### LINKED ISSUE #67865: Stateless test `02164_clickhouse_local_interactive_exception` is flaky (a bit)
Example

https://s3.amazonaws.com/clickhouse-test-reports/0/c0a3690cc8f01e5ef66ae22aa4122ae023e08828/stateless_tests__msan__[2_4].html

[CI database](https://play.clickhouse.com/play?user=play#U0VMRUNUIHB1bGxfcmVxdWVzdF9udW1iZXIsIGNoZWNrX3N0YXJ0X3RpbWUsIGNoZWNrX25hbWUsIHRlc3RfbmFtZSwgdGVzdF9zdGF0dXMsIGNoZWNrX3N0YXR1cywgcmVwb3J0X3VybApGUk9NIGNoZWNrcwpXSEVSRSAxCiAgICBBTkQgcHVsbF9yZXF1ZXN0X251bWJlciA9IDAKICAgIEFORCB0ZXN0X3N0YXR1cyAhPSAnU0tJUFBFRCcKICAgIEFORCB0ZXN0X3N0YXR1cyAhPSAnT0snCiAgICBBTkQgY2hlY2tfc3RhdHVzICE9ICdzdWNjZXNzJwogICAgQU5EIHRlc3RfbmFtZSBsaWtlICclMDIxNjRfY2xpY2tob3VzZV9sb2NhbF9pbnRlcmFjdGl2ZV9leGNlcHRpb24lJwogICAgQU5EIGNoZWNrX3N0YXJ0X3RpbWUgPiBub3coKSAtIGludGVydmFsIDIgbW9udGgKT1JERVIgQlkgY2hlY2tfc3RhcnRfdGltZSBkZXNjLCBjaGVja19uYW1lLCB0ZXN0X25hbWU=)

### PR DESCRIPTION
**Pull request by thevar1able**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixes https://github.com/ClickHouse/ClickHouse/issues/67865

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

File: tests/queries/0_stateless/02164_clickhouse_local_interactive_exception.python [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    if shell_pid == 0:
+        os.execv(program, argv)
+    else:
+        try:
+            debug_log_fd = open(DEBUG_LOG, "a")
+
+            output_b = os.read(master, 4096)
+            output = output_b.decode()
+            debug_log_fd.write(repr(output_b) + "\n")
+            debug_log_fd.flush()
```
Comment (evillique) [inline tests/queries/0_stateless/02164_clickhouse_local_interactive_exception.python:55]: It looks like this is reused a bunch of times in this test and can be moved into a separate function

File: tests/queries/0_stateless/02164_clickhouse_local_interactive_exception.python [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+                continue
+
+            print(f"Timeout, state: {STATE_MAP[state.value]}")
+            return
+
+def expect(text, master, debug_log_fd):
+    output_b = os.read(master, 4096)
+    output = output_b.decode()
+    debug_log_fd.write(repr(output_b) + "\n")
+    debug_log_fd.flush()
```
Comment (evillique) [inline tests/queries/0_stateless/02164_clickhouse_local_interactive_exception.python:47]:
```suggestion
def expect(text, master, debug_log_fd, output=""):
```

File: tests/queries/0_stateless/02164_clickhouse_local_interactive_exception.python [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+            os.write(master, b"\r")
+            state.value = 2
+
+            output = expect("CANNOT_PARSE_INPUT_ASSERTION_FAILED", master, debug_log_fd)
+
+            while not ":)" in output:
+                output_b = os.read(master, 4096)
+                output += output_b.decode()
+                debug_log_fd.write(repr(output_b) + "\n")
+                debug_log_fd.flush()
```
Comment (evillique) [inline tests/queries/0_stateless/02164_clickhouse_local_interactive_exception.python:80]: expect(":)", master, debug_log_fd, output)

Comment (evillique) [reply tests/queries/0_stateless/02164_clickhouse_local_interactive_exception.python:80]:
```suggestion
            expect(":)", master, debug_log_fd, output)
```

Comment (evillique) [inline tests/queries/0_stateless/02164_clickhouse_local_interactive_exception.python:80]:
```suggestion
            expect(":)", master, debug_log_fd, output)
```

**Comment by nikitamikhaylov** [comment]
WDYT of using [`pwntools`](https://github.com/Gallopsled/pwntools) for capturing the stdout/stderr of a process. I think it might be better

**Comment by thevar1able** [comment]
Yes, `[pwnlib.tubes.process](https://docs.pwntools.com/en/stable/tubes/processes.html#module-pwnlib.tubes.process)` looks like good match, however it is already implemented in the test with stdlib functionality, and `pwntools` is huge lib covering a ton of things. So I'd say I don't want it here, but feel free to do an improvement in a separate PR if you are sure it might be better.

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit 34d2c71eadfefd79e3cfb62d46d156cdbdaab681 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/68054/34d2c71eadfefd79e3cfb62d46d156cdbdaab681/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Unit tests</td><td>Runs the unit tests for different release types</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/42ae591017f410fffe5696ab396723f27626a528/unit_tests__release_.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Review by evillique — APPROVED** [review]

**Pull request merged**